app/src/main/java/ch/epfl/cs311/wanderwave/ui/screens/MapScreen.kt

```diff
@@ -1,79 +1,163 @@
+  return permissionState.permissions.any { !it.status.isGranted }
+}
 
+@OptIn(ExperimentalPermissionsApi::class)
+@SuppressLint("MissingPermission")
+@Composable
+fun MapScreen(navigationActions: NavigationActions, viewModel: MapViewModel = hiltViewModel()) {
   val context = LocalContext.current
-  val isMapReady = remember { mutableStateOf(false) }
-  val googleMapState = viewModel.googleMapState.collectAsState().value
-  // GoogleMap(modifier = Modifier.testTag("mapScreen")) { DisplayBeacons(uiState.beacons) }
+  val cameraPositionState: CameraPositionState = rememberCameraPositionState {}
+  val mapIsLoaded = remember { mutableStateOf(false) }
 
-  Box(modifier = Modifier.fillMaxSize().testTag("mapScreen"), contentAlignment = Alignment.Center) {
-    AndroidView(
-        modifier = Modifier.testTag("map"),
-        factory = { ctx ->
-          MapView(ctx).apply {
-            onCreate(Bundle())
-            onResume()
-            getMapAsync { googleMap ->
-              viewModel.setGoogleMap(googleMap)
-              isMapReady.value = true
-              googleMap.setMapStyle(MapStyleOptions.loadRawResourceStyle(context, R.raw.map_style))
-            }
-          }
-        },
-    )
+  val permissionState =
+      rememberMultiplePermissionsState(
+          listOf(
+              Manifest.permission.ACCESS_COARSE_LOCATION,
+              Manifest.permission.ACCESS_FINE_LOCATION,
+          ))
+
+  // remember the camera position when navigating away
+  DisposableEffect(Unit) {
+    onDispose { viewModel.cameraPosition.value = cameraPositionState.position }
+  }
+
+  // if we have permission, show the location, otherwise show the map without location
+  if (permissionState.allPermissionsGranted) {
+    WanderwaveGoogleMap(
+        cameraPositionState,
+        viewModel,
+        onMapLoaded = {
+          println("Map is loaded!")
+          mapIsLoaded.value = true
+        })
+  } else {
+    WanderwaveGoogleMap(cameraPositionState, viewModel)
+  }
 
-    if (!isMapReady.value) {
-      CircularProgressIndicator(modifier = Modifier.testTag("CircularProgressIndicator"))
-    } else {
-      googleMapState?.let { googleMap ->
-        val epfl = LatLng(46.518831258, 6.559331096)
-        googleMap.addMarker(MarkerOptions().position(epfl).title("Marker at EPFL"))
-        googleMap.moveCamera(CameraUpdateFactory.newLatLng(epfl))
+  if (needToRequestPermissions(permissionState)) {
+    AskForPermissions(permissionState)
+  } else {
+    // if we have location permissions, move the camera to the last known location **once**
+    val location = viewModel.getLastKnownLocation(context)
+    LaunchedEffect(location != null, mapIsLoaded.value) {
+      if (location != null && mapIsLoaded.value) {
+        moveCamera(cameraPositionState, location, viewModel.cameraPosition.value)
       }
     }
   }
 }
 
+@Composable
+fun WanderwaveGoogleMap(cameraPositionState: CameraPositionState, viewModel: MapViewModel) {
```

Comment:
can't you prevent copying of code by calling this function in the other overloaded function ? or vice versa ? 
`fun WanderwaveGoogleMap(cameraPositionState: CameraPositionState, viewModel: MapViewModel) {
WanderwaveGoogleMap(
    cameraPositionState: CameraPositionState = cameraPositionState,
    viewModel: MapViewModel = viewModel,
    onMapLoaded: () -> Unit = {},
)
`

Reply:
I agree that the duplication is a bit unfortunate... the problem is that only the second implementation (with the onMapLoaded function) should set isMyLocationEnabled to true.

I guess it's better to only have one function and pass isLocationEnabled as an argument as well.